### PR TITLE
Add a jupyter lab workspace

### PR DESCRIPTION
**Pull request by jjalvare**

### CODE REVIEW

File: terraform/jupyterlab-ibmcloud/README.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
+# Deploy a Jupyter Lab Workspace on IBM Cloud VPC 
+
+Deploy a Juptyer Lab on Ubuntu 18.04 ppc64le running in IBM Cloud VPC
```
Comment (jwcroppe) [inline terraform/jupyterlab-ibmcloud/README.md]: Jupyter

File: terraform/jupyterlab-ibmcloud/README.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+1. Terraform >= 0.11.14 
+2. IBM Cloud Terraform Provider >= 0.20.0 
+3. IBM Cloud API Key with authorization to provision in IBM Cloud VPC 
+
+## Overview
+
+This deployment:
+  1. Creates a Linux Ubuntu ppc64le VM server on IBM Cloud VPC 
+  2. Creates a new ssh key to login 
+  3. Opens Ports 80, 443 in VPC Security Group port to access the Jupter Lab Workspace
```
Comment (jwcroppe) [inline terraform/jupyterlab-ibmcloud/README.md]: Jupyter

File: terraform/jupyterlab-ibmcloud/README.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+2. IBM Cloud Terraform Provider >= 0.20.0 
+3. IBM Cloud API Key with authorization to provision in IBM Cloud VPC 
+
+## Overview
+
+This deployment:
+  1. Creates a Linux Ubuntu ppc64le VM server on IBM Cloud VPC 
+  2. Creates a new ssh key to login 
+  3. Opens Ports 80, 443 in VPC Security Group port to access the Jupter Lab Workspace 
+  4. Opens Ports 22 to access SSH console
```
Comment (jwcroppe) [inline terraform/jupyterlab-ibmcloud/README.md]: port

File: terraform/jupyterlab-ibmcloud/README.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  5. Sets up GPU Driver 418.39 and starts services
+  6. Loads CUDA 10.1 Libraries to interact in docker environment
+  7. Downloads jupyter lab from docker hub: jjalvare/tensorflow-gpu-jupyter
+  8. Run jjalvare/tensorflow-gpu-jupyter as tensorflow-gpu-jupyter
+
+### Install Complete 
+After install you can access the default Jupyter Lab Web page of Ubuntu VM by IP:
```
Comment (jwcroppe) [inline terraform/jupyterlab-ibmcloud/README.md]: web

File: terraform/jupyterlab-ibmcloud/README.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+## Overview
+
+This deployment:
+  1. Creates a Linux Ubuntu ppc64le VM server on IBM Cloud VPC 
+  2. Creates a new ssh key to login 
+  3. Opens Ports 80, 443 in VPC Security Group port to access the Jupter Lab Workspace 
+  4. Opens Ports 22 to access SSH console
+  5. Sets up GPU Driver 418.39 and starts services
+  6. Loads CUDA 10.1 Libraries to interact in docker environment
+  7. Downloads jupyter lab from docker hub: jjalvare/tensorflow-gpu-jupyter
```
Comment (jwcroppe) [inline terraform/jupyterlab-ibmcloud/README.md]: Jupyter; Docker Hub

File: terraform/jupyterlab-ibmcloud/README.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  8. Run jjalvare/tensorflow-gpu-jupyter as tensorflow-gpu-jupyter
+
+### Install Complete 
+After install you can access the default Jupyter Lab Web page of Ubuntu VM by IP:
+
+    http://<IP>/?token=<token> 
+
+The token can be found in your output of terraform:
```
Comment (jwcroppe) [inline terraform/jupyterlab-ibmcloud/README.md]: Terraform

File: terraform/jupyterlab-ibmcloud/README.md (hunk 1 above)
Comment (jwcroppe) [inline terraform/jupyterlab-ibmcloud/README.md]: We should put another few sentences about what high-level use cases this enables for users that aren't necessarily familiar with Jupyter... i.e., a few segments from https://jupyterlab.readthedocs.io/en/stable/getting_started/overview.html#overview

File: terraform/jupyterlab-ibmcloud/scripts/install_common.sh [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    python3-dev python3-pip python3 git gnupg2 pass build-essential 
+    
+echo "Setup Python 3"
+update-alternatives --install /usr/bin/python python /usr/bin/python3 1 && update-alternatives --install /usr/bin/pip pip /usr/bin/pip3 1
+echo "Nvidia Runtime Repository"
+curl -s -L https://nvidia.github.io/nvidia-container-runtime/gpgkey | sudo apt-key add -
+distribution=$(. /etc/os-release;echo $ID$VERSION_ID) && curl -s -L https://nvidia.github.io/nvidia-container-runtime/$distribution/nvidia-container-runtime.list | sudo tee /etc/apt/sources.list.d/nvidia-container-runtime.list
+echo "Add docker repository"
+curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
+add-apt-repository "deb [arch=ppc64el] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
```
Comment (jwcroppe) [inline terraform/jupyterlab-ibmcloud/scripts/install_common.sh]: Do we need this?  Might be needed for Docker, but just confirming.

File: terraform/jupyterlab-ibmcloud/scripts/install_docker.sh [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash -xe
+curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
+apt-key fingerprint 0EBFCD88
+arch=`dpkg --print-architecture`
+add-apt-repository "deb [arch=${arch}] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
```
Comment (jwcroppe) [inline terraform/jupyterlab-ibmcloud/scripts/install_docker.sh:9]: Ditto?

File: terraform/jupyterlab-ibmcloud/scripts/install_cuda.sh [hunk 9]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash -xe
+# Install cuda driver and libraries
```
Comment (jwcroppe) [inline terraform/jupyterlab-ibmcloud/scripts/install_cuda.sh]: CUDA; and would it make sense to have this get the latest version, or don't they have convenient links for the 'newest' driver?  Or was there a reason you needed this version?  Just thinking of trying to keep it 'current' over time with minimal updates.

File: terraform/jupyterlab-ibmcloud/scripts/install_runtime.sh [hunk 10]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash -xe
+# Install nvidia runtim in daemon.json and reload docker
```
Comment (jwcroppe) [inline terraform/jupyterlab-ibmcloud/scripts/install_runtime.sh:2]: runtime

**Review by jwcroppe — CHANGES_REQUESTED** [review]
Nice additional - a few comments below to address

File: terraform/jupyterlab-ibmcloud/scripts/install_docker.sh (hunk 8 above)
Comment (jjalvare) [reply terraform/jupyterlab-ibmcloud/scripts/install_docker.sh:9]: thanks. i must of forgotten this.

File: terraform/jupyterlab-ibmcloud/README.md (hunk 1 above)
Comment (jjalvare) [reply terraform/jupyterlab-ibmcloud/README.md]: done

File: terraform/jupyterlab-ibmcloud/scripts/install_cuda.sh (hunk 9 above)
Comment (jjalvare) [reply terraform/jupyterlab-ibmcloud/scripts/install_cuda.sh]: I am running tensorflow 1.15 in the Jupyter Lab and wanted to stick with a known workin Cuda. I agree, needs to be less costly in updates. Will add a todo to make it more resilient to cuda library changes for now.

**Pull request merged**